R/textmodel_svmlin.R

```diff
@@ -189,3 +208,107 @@ coefficients.textmodel_svmlin <- function(object, ...) {
 print.predict.textmodel_svmlin <- function(x, ...) {
     print(unclass(x))
 }
+
+
+
+# adapted from RSSL -----------------
+
+svmlin <- function(X, y, X_u = NULL, algorithm = 1, lambda = 1, lambda_u = 1, max_switch = 10000,
+                   pos_frac = 0.5, Cp = 1.0, Cn = 1.0, verbose = FALSE,
+                   intercept = TRUE, scale = FALSE, x_center = FALSE) {
+
+  use_Xu_for_scaling <- TRUE
+  classnames <- levels(y)
+
+    if (scale | x_center) {
+        if (intercept) {
+            cols <- 2:ncol(X) # do not scale the intercept column
+        } else {
+            cols <- 1:ncol(X)
+        }
+
+        if (!is.null(X_u) && use_Xu_for_scaling) {
+            Xe <- rbind(X, X_u)
+            scaling <- scaleMatrix(Xe[, cols, drop = FALSE], center = TRUE, scale = scale)
+            X[, cols] <- predict(scaling, as.matrix(X[, cols, drop=FALSE]))
+            X_u[, cols] <- predict(scaling, as.matrix(X_u[, cols, drop=FALSE]))
+        } else {
+            scaling <- scaleMatrix(X[, cols, drop = FALSE], center = TRUE, scale = scale)
+            X[, cols] <- predict(scaling, as.matrix(X[, cols, drop=FALSE]))
+            if (!is.null(X_u)) {
+                X_u[, cols] <- predict(scaling,as.matrix(X_u[, cols, drop=FALSE]))
+            }
+        }
+    } else {
+        scaling = NULL
+    }
+
+    y <- as.numeric(y) * 2 - 3
+
+    # Combine feature matrices, add intercept and transpose them to conform to the C++ datastructure
+    if (is.null(X_u) || algorithm < 2) {
+        if (intercept) {
+            X <- cbind(X, 1)
+            X_u <- cbind(X_u, 1)
+        }
+        Xall <- Matrix::t(X)
+    } else {
+        if (intercept) {
+            X <- cbind(X, 1)
+            X_u <- cbind(X_u, 1)
+        }
+        Xall <- Matrix::t(Matrix::rbind2(X, X_u))
+        y <- c(y, rep(0,nrow(X_u)))
+    }
+
+    # Determine costs
+    costs <- rep(1, ncol(Xall))
+    if (algorithm < 1) {
+        costs[y < 0] <- Cn
+        costs[y > 0] <- Cp
+    }
+
+    res <- svmlin_rcpp(X = Xall, y = y, l = nrow(X), algorithm = algorithm,
+                       lambda = lambda, lambda_u = lambda_u, max_switch = max_switch,
+                       pos_frac = pos_frac, Cp = Cp, Cn = Cn, costs = costs,
+                       verbose = verbose)
+
+    list(classnames = classnames,
+         weights = res$Weights,
+         algorithm = algorithm,
+         scaling = scaling,
+         intercept = intercept)
+}
+
+setClass("scaleMatrix",
```

Comment:
Do we need really this object? Looks related to `x_u`.

Reply:
see above